src/Appwrite/Auth/OAuth2/Auth0.php

```diff
@@ -0,0 +1,207 @@
+<?php
+
+namespace Appwrite\Auth\OAuth2;
+
+use Appwrite\Auth\OAuth2;
+
+// Reference Material
+// https://auth0.com/docs/api/authentication
+
+class Auth0 extends OAuth2
+{  
+     /**
+     * @var array
+     */
+    protected $scopes = [
+        'openid',
+        'profile',
+        'email',
+        'offline_access'
+    ];
+    
+    /**
+     * @var array
+     */
+    protected $user = [];
+    
+    /**
+     * @var array
+     */
+    protected $tokens = [];
+    
+    /**
+     * @return string
+     */
+    public function getName(): string
+    {
+        return 'auth0';
+    }
+
+    /**
+     * @return string
+     */
+    public function getLoginURL(): string
+    {
+        return 'https://'.$this->getAuth0Domain().'/authorize?'.\http_build_query([
+            'client_id' => $this->appID,
+            'redirect_uri' => $this->callback,
+            'state'=> \json_encode($this->state),
+            'scope'=> \implode(' ', $this->getScopes()),
+            'response_type' => 'code'
+        ]);
+    }
+
+    /**
+     * @param string $code
+     *
+     * @return array
+     */
+    protected function getTokens(string $code): array
+    {
+        if(empty($this->tokens)) {
+            $headers = ['Content-Type: application/x-www-form-urlencoded'];
+            $this->tokens = \json_decode($this->request(
+                'POST',
+                'https://'.$this->getAuth0Domain().'/oauth/token',
+                $headers,
+                \http_build_query([
+                    'code' => $code,
+                    'client_id' => $this->appID,
+                    'client_secret' => $this->getClientSecret(),
+                    'redirect_uri' => $this->callback,
+                    'scope' => \implode(' ', $this->getScopes()),
+                    'grant_type' => 'authorization_code'
+                ])
+            ), true);
+        }
+
+        return $this->tokens;
+    }
+    
+    
+    /**
+     * @param string $refreshToken
+     *
+     * @return array
+     */
+    public function refreshTokens(string $refreshToken):array
```

Comment:
```suggestion
    public function refreshTokens(string $refreshToken): array
```
PSR-12 style guides
<img width="1143" alt="Screenshot 2022-04-02 at 4 18 50 PM" src="https://user-images.githubusercontent.com/20852629/161385135-6ea771e7-81db-421a-a6a6-9cbfadf550ef.png">